plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+
+ where
+  -- | Prior to starting the second pass, we clean the term from all its let-declarations and store them separately in a table.
+  -- The 2nd pass will later re-introduce these let-declarations, potentially placing them differently than before, thus essentially "floating the lets".
+  (pirClean :: Term tyname name uni a, rhsTable :: RhsTable tyname name uni a) = removeLets pir
+
+  -- 2nd-pass functions
+  ---------------------
+
+  -- | visit each term to apply the float transformation
+  -- OPTIMIZE: stop descending when state==mempty?
+  goTerm :: Int -- ^ current depth
+         -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+         -> Term tyname name uni a
+         -> State [S.Set PLC.Unique] (Term tyname name uni a)
+  goTerm curDepth searchTable = \case
+      -- we are only interested in lambdas/Lambdas
+      LamAbs a n ty tBody -> LamAbs a n ty <$> goAbs curDepth searchTable n tBody
+      TyAbs a n k tBody  -> TyAbs a n k <$> goAbs curDepth searchTable n tBody
+      -- descend otherwise to apply the transformations to subterms
+      t -> termSubterms (goTerm curDepth searchTable) t
+
+  -- | If a lambda/Lambda is found, the current location is updated (depth+lamUnique) and try to float its body
+  goAbs :: PLC.HasUnique b b'
+        => Int -- ^ current depth
+        -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+        -> b                  -- ^ lambda/Lambda's unique
+        -> Term tyname name uni a -- ^ lambda/Lambda's body
+        -> State [S.Set PLC.Unique] (Term tyname name uni a)
+  goAbs oldDepth floatData n tBody =
+    let newLam = LamDep (oldDepth+1) (n^.PLC.theUnique)
+    in goFloat newLam tBody floatData
+
+  -- | We are currently INSIDE (exactly under) a lambda/Lambda body/Top (Top if we are right at the start of the algorithm)
+  -- We try to see if we have some lets to float here based on our 1st-pass-table data (searchTable).
+  goFloat :: Rank -- ^ the rank/location of the lambda above that has this body
+         -> Term tyname name uni a                           -- ^ the body term
+         -> FloatData -- ^ the lambdas we are searching for (to float some lets inside them)
+         -> State [S.Set PLC.Unique] (Term tyname name uni a)                           -- ^ the transformed body term
+  goFloat aboveLamLoc tBody floatData  =
+   -- look for the next smallest depth remaining to place
+   case IM.minViewWithKey floatData of
+     Nothing -> pure tBody -- nothing left to float
+     Just ((searchingForDepth, searchingForLams_Lets), restFloatData) ->
+      let curDepth = aboveLamLoc^.depth
+      in case curDepth `compare` searchingForDepth of
+      -- the minimum next depth we are looking for, is not this one, so just descend with the whole floatdata
+      LT -> goTerm curDepth floatData tBody
```

Comment:
Suboptimal.

It seems that instead of using `FloatData` as state (well, technically, context/environment), we should split it into (braces are for clarity) {the minimal key and stuff associated with it} and the rest of `FloatData` outside of that function. Then we don't need to retrieve that minimal key and what is associated with it every time we encounter a binder -- only when we encounter a binder that we have some `let`s to merge under. And we can also stop early once the `FloatData` becomes empty.

Reply:
It also seems that we don't need `IntMap` here as we always access it in a "consecutive" fashion. It would be more convenient to simply turn the `IntMap` into a list and use that list in the second pass.

Reply:
Y it was @michaelpj suggestion to keep the intmap structure to enforce the sorting. It is slightly slower this way though.